tbl_custom_summary()

larmarange opened the pull request:
- New function `tbl_custom_summary()` allowing to create a table of summary statistics using a custom summary function (#973, #976)
- Set of helpers to be used with `tbl_custom_summary()`: `continuous_summary()`, `proportion_summary()`, `ratio_summary()


closes #973

First draft of a tbl\_custom\_summary() function to test the concept
Instead of building an empty table and playing with modify\_table\_body(),
we started with the current code of tbl\_summary(), changing the way df\_stats
is computed within meta\_data, and then using all the existing mechanisms
of tbl\_summary() to generate the table.

The heart of tbl\_custom\_summary() is that users have to provide through the
stat\_fns argument a custom function that will compute the stats, and they
specify in the statistic argument how to display the statistics using the
same syntax as in tbl\_summary().

The stat\_fns argument specify the custom function(s) to be used for computing
the summary statistics. For example, `stat_fns = everything() ~ foo`.

Each function may take the following arguments:
`foo(data, full_data, variable, by, type, ...)`
- `data=` is the input data frame passed to `tbl_summary()`, subsetted
according to the level of `by` or `variable` if any, excluding `NA`
values of the current `variable`
- `full_data=` is the full input data frame passed to `tbl_summary()`
- `variable=` is a string indicating the variable to perform the
calculation on
- `by=` is a string indicating the by variable from `tbl_summary=`,
if present
- `type=` is a string indicating the type of variable
(continous, categorical, …)

The user-defined does not need to utilize each of these inputs. It’s
encouraged the user-defined function accept `...` as each of the arguments
*will* be passed to the function, even if not all inputs are utilized by
the user’s function, e.g. `foo(data, ..)`

One or two additional columns are added to `data` and `full_data`:
- “.variable”, a copy of the current variable defined in `variable`
- “.by”, a copy of the by variable defined in `by`, if any

### A FIRST EXAMPLE

``` r
library(gtsummary)

mean_age <- function(data,...) {
  dplyr::tibble(mean_age = mean(data$age, na.rm = TRUE))
}
mean_age(trial)
#> # A tibble: 1 x 1
#>   mean_age
#>      <dbl>
#> 1     47.2

trial %>%
  tbl_custom_summary(
    include = c("grade", "response", "stage"),
    by = "trt",
    stat_fns = everything() ~ mean_age,
    statistic = everything() ~ "{mean_age}",
    digits = everything() ~ 1
  ) %>%
  add_overall(last = TRUE) %>%
  modify_footnote(all_stat_cols() ~ "Mean age") %>%
  as_kable()
```

| **Characteristic** | **Drug A**, N = 98 | **Drug B**, N = 102 | **Overall**, N = 200 |
|:-------------------|:-------------------|:--------------------|:---------------------|
| Grade              |                    |                     |                      |
| I                  | 45.9               | 46.4                | 46.2                 |
| II                 | 44.6               | 50.3                | 47.5                 |
| III                | 51.0               | 45.7                | 48.1                 |
| Tumor Response     | 50.1               | 49.5                | 49.8                 |
| Unknown            | 3                  | 4                   | 7                    |
| T Stage            |                    |                     |                      |
| T1                 | 44.1               | 49.5                | 46.8                 |
| T2                 | 50.2               | 46.4                | 48.1                 |
| T3                 | 48.8               | 50.0                | 49.4                 |
| T4                 | 45.3               | 44.3                | 44.8                 |

To facililitate the writing of the functions, users don’t have to perform
the subset by themselves as data is already subsetted according to
variable level and by level if any. However, the full dataset is also passed
to the function.

``` r
diff_to_great_mean <- function(data, full_data, ...) {
  mean <- mean(data$marker, na.rm = TRUE)
  great_mean <- mean(full_data$marker, na.rm = TRUE)
  diff <- mean - great_mean
  dplyr::tibble(
    mean = mean,
    great_mean = great_mean,
    diff = diff
  )
}

trial %>%
  tbl_custom_summary(
    include = c("grade", "stage"),
    by = "trt",
    stat_fns = everything() ~ diff_to_great_mean,
    statistic = everything() ~ "{mean} (diff: {diff})",
    digits = everything() ~ c(2, 2)
  ) %>%
  add_overall(last = TRUE) %>%
  modify_footnote(all_stat_cols() ~ "Mean of marker (diff to great mean)") %>%
  as_kable()
```

| **Characteristic** | **Drug A**, N = 98 | **Drug B**, N = 102 | **Overall**, N = 200 |
|:-------------------|:-------------------|:--------------------|:---------------------|
| Grade              |                    |                     |                      |
| I                  | 1.15 (diff: 0.23)  | 0.98 (diff: 0.06)   | 1.07 (diff: 0.15)    |
| II                 | 0.86 (diff: -0.06) | 0.54 (diff: -0.38)  | 0.68 (diff: -0.24)   |
| III                | 1.02 (diff: 0.10)  | 0.98 (diff: 0.06)   | 1.00 (diff: 0.08)    |
| T Stage            |                    |                     |                      |
| T1                 | 0.74 (diff: -0.17) | 0.66 (diff: -0.25)  | 0.70 (diff: -0.21)   |
| T2                 | 1.22 (diff: 0.30)  | 1.01 (diff: 0.10)   | 1.10 (diff: 0.19)    |
| T3                 | 1.06 (diff: 0.15)  | 0.94 (diff: 0.02)   | 1.00 (diff: 0.09)    |
| T4                 | 1.08 (diff: 0.17)  | 0.67 (diff: -0.25)  | 0.85 (diff: -0.06)   |

To simplify the writing, in particular for users who are not familial with \[\[\]\],
we also provide a .variable column, i.e. data$.variable == data\[\[variable\]\].

``` r
geom <- function(data, ...) {
  dplyr::tibble(
    geom_mean = exp(mean(log(data$.variable), na.rm = TRUE))
  )
}

trial %>%
  tbl_custom_summary(
    include = c("age", "marker", "ttdeath"),
    by = "grade",
    stat_fns = everything() ~ geom,
    statistic = everything() ~ "{geom_mean}"
  ) %>%
  as_kable()
```

| **Characteristic**     | **I**, N = 68 | **II**, N = 68 | **III**, N = 64 |
|:-----------------------|:--------------|:---------------|:----------------|
| Age                    | 43            | 45             | 46              |
| Unknown                | 2             | 6              | 3               |
| Marker Level (ng/mL)   | 0.60          | 0.34           | 0.58            |
| Unknown                | 2             | 5              | 3               |
| Months to Death/Censor | 20.4          | 18.0           | 17.7            |

It is also possible to write some generic functions that could be provided
within gtsummary. For example, for a ratio table:

``` r
ratio_summary <- function(numerator, denominator, na.rm = TRUE) {
  function(data, ...) {
    num <- sum(data[[numerator]], na.rm = na.rm)
    denom <- sum(data[[denominator]], na.rm = na.rm)
    ratio <- num / denom
    if (num %% 1 == 0) {
      ci_poisson <- poisson.test(num, denom)$conf.int
    } else {
      ci_poisson <- c(NA, NA)
    }
    dplyr::tibble(
      num = num,
      denom = denom,
      ratio = ratio,
      conf.poisson.low = ci_poisson[1],
      conf.poisson.high = ci_poisson[2]
    )
  }
}

trial %>%
  tbl_custom_summary(
    stat_fns = everything() ~ ratio_summary("response", "ttdeath"),
    statistic = everything() ~ "{ratio} [{conf.poisson.low}; {conf.poisson.high}] ({num}/{denom})",
    digits = everything() ~ c(3, 2, 2, 0, 0),
    by = "trt",
    include = c("stage", "grade")
  ) %>%
  add_overall(last = TRUE) %>%
  bold_labels() %>%
  modify_footnote(
    update = all_stat_cols() ~ "Incidence [95 CI] (cases / observation time)"
  ) %>%
  as_kable()
```

| **Characteristic** | **Drug A**, N = 98            | **Drug B**, N = 102           | **Overall**, N = 200            |
|:-------------------|:------------------------------|:------------------------------|:--------------------------------|
| **T Stage**        |                               |                               |                                 |
| T1                 | 0.012 \[0.00; 0.02\] (7/583)  | 0.021 \[0.01; 0.04\] (11/522) | 0.016 \[0.01; 0.03\] (18/1,105) |
| T2                 | 0.011 \[0.00; 0.02\] (6/528)  | 0.012 \[0.01; 0.03\] (7/560)  | 0.012 \[0.01; 0.02\] (13/1,089) |
| T3                 | 0.019 \[0.01; 0.04\] (8/426)  | 0.016 \[0.01; 0.03\] (7/425)  | 0.018 \[0.01; 0.03\] (15/852)   |
| T4                 | 0.016 \[0.01; 0.03\] (7/445)  | 0.018 \[0.01; 0.04\] (8/434)  | 0.017 \[0.01; 0.03\] (15/879)   |
| **Grade**          |                               |                               |                                 |
| I                  | 0.011 \[0.00; 0.02\] (8/734)  | 0.019 \[0.01; 0.03\] (13/690) | 0.015 \[0.01; 0.02\] (21/1,424) |
| II                 | 0.011 \[0.00; 0.02\] (7/651)  | 0.019 \[0.01; 0.03\] (12/645) | 0.015 \[0.01; 0.02\] (19/1,296) |
| III                | 0.022 \[0.01; 0.04\] (13/598) | 0.013 \[0.01; 0.03\] (8/607)  | 0.017 \[0.01; 0.03\] (21/1,204) |

### ADDITIONAL COMMENTS

The default digits works quite well because `adding_formatting_as_attr()`
already default to `style_number`. However, it does not work well if the stat
is not numeric. Maybe relevant to test if the stat is numeric, and if not to
default to `as.character()`? Could be done in a separate PR.
In the following example, if digits is not specified, it will produce an error.

``` r
foo <- function(data, full_data, ...) {
  mean <- mean(data$.variable, na.rm = TRUE)
  great_mean <- mean(full_data$.variable, na.rm = TRUE)
  dplyr::tibble(
    mean = mean,
    great_mean = great_mean,
    pos = ifelse(mean > great_mean, "high", "low")
  )
}

trial %>%
  tbl_custom_summary(
    include = c("age", "marker", "ttdeath"),
    by = "grade",
    stat_fns = everything() ~ foo,
    statistic = everything() ~ "{mean} ({pos})",
    digits = everything() ~ c(style_number, as.character)
  ) %>%
  as_kable()
```

| **Characteristic**     | **I**, N = 68 | **II**, N = 68 | **III**, N = 64 |
|:-----------------------|:--------------|:---------------|:----------------|
| Age                    | 46 (low)      | 48 (high)      | 48 (high)       |
| Unknown                | 2             | 6              | 3               |
| Marker Level (ng/mL)   | 1 (high)      | 1 (low)        | 1 (high)        |
| Unknown                | 2             | 5              | 3               |
| Months to Death/Censor | 21 (high)     | 19 (low)       | 19 (low)        |

### Can we simplify the syntax?

For many arguments in gtsummary, formulas or list of formulas should be
provided, and each formula should have a left side.
Could we consider that `~ something` could be interpreted as
`everything() ~ something`?
It could be common to all gtsummary functions and discussed in a seperate PR,
and implemented through `.formula_list_to_named_list()`

### Overall row?

There is an easy trick to have an overall row by creating a variable containing
only TRUE. Should it be an option to add to the function?

``` r
trial$Overall <- TRUE

trial %>%
  tbl_custom_summary(
    stat_fns = everything() ~ ratio_summary("response", "ttdeath"),
    statistic = everything() ~ "{ratio} [{conf.poisson.low}; {conf.poisson.high}] ({num}/{denom})",
    digits = everything() ~ c(3, 2, 2, 0, 0),
    by = "trt",
    include = c("stage", "grade", "Overall")
  ) %>%
  add_overall(last = TRUE) %>%
  bold_labels() %>%
  modify_footnote(
    update = all_stat_cols() ~ "Incidence [95 CI] (cases / observation time)"
  ) %>%
  as_kable()
```

| **Characteristic** | **Drug A**, N = 98              | **Drug B**, N = 102             | **Overall**, N = 200            |
|:-------------------|:--------------------------------|:--------------------------------|:--------------------------------|
| **T Stage**        |                                 |                                 |                                 |
| T1                 | 0.012 \[0.00; 0.02\] (7/583)    | 0.021 \[0.01; 0.04\] (11/522)   | 0.016 \[0.01; 0.03\] (18/1,105) |
| T2                 | 0.011 \[0.00; 0.02\] (6/528)    | 0.012 \[0.01; 0.03\] (7/560)    | 0.012 \[0.01; 0.02\] (13/1,089) |
| T3                 | 0.019 \[0.01; 0.04\] (8/426)    | 0.016 \[0.01; 0.03\] (7/425)    | 0.018 \[0.01; 0.03\] (15/852)   |
| T4                 | 0.016 \[0.01; 0.03\] (7/445)    | 0.018 \[0.01; 0.04\] (8/434)    | 0.017 \[0.01; 0.03\] (15/879)   |
| **Grade**          |                                 |                                 |                                 |
| I                  | 0.011 \[0.00; 0.02\] (8/734)    | 0.019 \[0.01; 0.03\] (13/690)   | 0.015 \[0.01; 0.02\] (21/1,424) |
| II                 | 0.011 \[0.00; 0.02\] (7/651)    | 0.019 \[0.01; 0.03\] (12/645)   | 0.015 \[0.01; 0.02\] (19/1,296) |
| III                | 0.022 \[0.01; 0.04\] (13/598)   | 0.013 \[0.01; 0.03\] (8/607)    | 0.017 \[0.01; 0.03\] (21/1,204) |
| **Overall**        | 0.014 \[0.01; 0.02\] (28/1,983) | 0.017 \[0.01; 0.02\] (33/1,942) | 0.016 \[0.01; 0.02\] (61/3,925) |

<sup>Created on 2021-09-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

larmarange commented:
In theory, it should be easy to reproduce `tbl_continuous()`. See code below.

However, I have an unexpected error.

``` r
library(gtsummary)
#> #Uighur

continuous_summary <- function(variable_to_summarize) {
  function(data, stat_display, ...) {
    gtsummary:::summarize_continuous(
      data,
      variable = variable_to_summarize,
      by = NULL, 
      stat_display = stat_display,
      summary_type = "continuous"
    ) %>%
      dplyr::select(-.data$variable, -.data$stat_display)
  }
}

continuous_summary("age")(trial, "{mean} {p15}")
#> # A tibble: 1 x 2
#>    mean   p15
#>   <dbl> <dbl>
#> 1  47.2  33.2
continuous_summary("age")(trial %>% dplyr::filter(trt == "Drug A"), "{mean} {p15}")
#> # A tibble: 1 x 2
#>    mean   p15
#>   <dbl> <dbl>
#> 1  47.0  32.5
continuous_summary("age")(trial %>% dplyr::filter(trt == "Drug B"), "{mean} {p15}")
#> # A tibble: 1 x 2
#>    mean   p15
#>   <dbl> <dbl>
#> 1  47.4    34

trial %>% dplyr::summarise(continuous_summary("age")(.data, "{mean} {p15}"))
#> Error: Problem with `summarise()` input `..1`.
#> i `..1 = continuous_summary("age")(.data, "{mean} {p15}")`.
#> x pas de méthode pour 'select' applicable pour un objet de classe "rlang_data_pronoun"
trial %>% 
  group_by(trt) %>%
  dplyr::summarise(continuous_summary("age")(.data, "{mean} {p15}"))
#> Error in group_by(., trt): impossible de trouver la fonction "group_by"

trial %>%
  tbl_custom_summary(
    include = c("trt"),
    stat_fns = everything() ~ continuous_summary("age"),
    statistic = everything() ~ "{mean} ({sd}) {p15}"
  )
#> Error: Problem with `mutate()` column `df_stats`.
#> i `df_stats = pmap(...)`.
#> x Problem with `summarise()` input `..2`.
#> i `..2 = stat_fn(...)`.
#> x pas de méthode pour 'select' applicable pour un objet de classe "rlang_data_pronoun"
#> i The error occurred in group 1: .variable = "Drug A".
#> i The error occurred in group 1: .variable = "Drug A".
```

<sup>Created on 2021-09-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

larmarange commented:
It is also very easy to reproduce the behavior of your new `tbl_continuous()`, with the advantage of being far more generic.

Of course, the proposed `continuous_summary()` could be exported directly by `gtsummary`.

``` r
library(gtsummary)

continuous_summary <- function(variable_to_summarize) {
  function(data, stat_display, ...) {
    gtsummary:::summarize_continuous(
      data,
      variable = variable_to_summarize,
      by = NULL, 
      stat_display = stat_display,
      summary_type = "continuous"
    ) %>%
      dplyr::select(-.data$variable, -.data$stat_display)
  }
}


trial %>% 
  dplyr::group_by(trt, grade) %>%
  dplyr::group_modify(
    continuous_summary("age"),
    stat_display = "{median} ({p25}, {p75})"
  )
#> # A tibble: 6 x 5
#> # Groups:   trt, grade [6]
#>   trt    grade median   p25   p75
#>   <chr>  <fct>  <dbl> <dbl> <dbl>
#> 1 Drug A I       46    36    60  
#> 2 Drug A II      44.5  31    54.5
#> 3 Drug A III     51.5  42.2  60.2
#> 4 Drug B I       48    42    55  
#> 5 Drug B II      50.5  43    57.2
#> 6 Drug B III     45    36    52
trial %>%
  tbl_custom_summary(
    include = grade,
    by = trt,
    stat_fns = everything() ~ continuous_summary("age"),
    statistic = everything() ~ "{median} ({p25}, {p75})"
  ) %>%
  add_overall() %>%
  as_kable()
```

| **Characteristic** | **Overall**, N = 200 | **Drug A**, N = 98 | **Drug B**, N = 102 |
|:-------------------|:---------------------|:-------------------|:--------------------|
| Grade              |                      |                    |                     |
| I                  | 47 (37, 56)          | 46 (36, 60)        | 48 (42, 55)         |
| II                 | 48 (37, 57)          | 44 (31, 54)        | 50 (43, 57)         |
| III                | 47 (38, 58)          | 52 (42, 60)        | 45 (36, 52)         |

<sup>Created on 2021-09-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

larmarange commented:
And an example of a `proportion_summary()` function.

``` r
library(gtsummary)

proportion_summary <- function(variable, value, weights = NULL, na.rm = TRUE, conf.level = 0.95) {
  # ADD A METHOD TO HARMONIZE WITH add_ci()
  variable_to_summarize <- variable
  function(data, ...) {
    if (is.null(weights)) {
      n <- sum(data[[variable_to_summarize]] == value, na.rm = na.rm)
      N <- sum(!is.na(data[[variable_to_summarize]]), na.rm = na.rm)
    } else {
      n <- sum((data[[variable_to_summarize]] == value) * data[[weights]], na.rm = na.rm)
      N <- sum((!is.na(data[[variable_to_summarize]])) * data[[weights]], na.rm = na.rm)
    }
    if (anyNA(n, N)) {
      ci <- c(NA, NA)
    } else {
      ci <- prop.test(n, N, conf.level = conf.level)$conf.int
    }
    dplyr::tibble(
      n = n,
      N = N,
      prop = n/N,
      conf.low = ci[1],
      conf.high = ci[2]
    )
  }
}

Titanic %>%
  dplyr::as_tibble() %>%
  dplyr::mutate(Overall = TRUE) %>%
  tbl_custom_summary(
    include = c("Age", "Class", "Overall"),
    by = "Sex",
    stat_fns = everything() ~ proportion_summary("Survived", "Yes", weights = "n"),
    statistic = everything() ~ "{prop}% ({n}/{N}) [{conf.low}-{conf.high}]",
    digits = everything() ~ list(function(x) {style_percent(x, digits = 1)}, 1, 1, style_percent, style_percent)
  ) %>%
  add_overall(last = TRUE) %>%
  as_kable()
```

| **Characteristic** | **Female**, N = 16            | **Male**, N = 16                | **Overall**, N = 32             |
|:-------------------|:------------------------------|:--------------------------------|:--------------------------------|
| Age                |                               |                                 |                                 |
| Adult              | 74.4% (316.0/425.0) \[70-78\] | 20.3% (338.0/1,667.0) \[18-22\] | 31.3% (654.0/2,092.0) \[29-33\] |
| Child              | 62.2% (28.0/45.0) \[47-76\]   | 45.3% (29.0/64.0) \[33-58\]     | 52.3% (57.0/109.0) \[43-62\]    |
| Class              |                               |                                 |                                 |
| 1st                | 97.2% (141.0/145.0) \[93-99\] | 34.4% (62.0/180.0) \[28-42\]    | 62.5% (203.0/325.0) \[57-68\]   |
| 2nd                | 87.7% (93.0/106.0) \[80-93\]  | 14.0% (25.0/179.0) \[9.4-20\]   | 41.4% (118.0/285.0) \[36-47\]   |
| 3rd                | 45.9% (90.0/196.0) \[39-53\]  | 17.3% (88.0/510.0) \[14-21\]    | 25.2% (178.0/706.0) \[22-29\]   |
| Crew               | 87.0% (20.0/23.0) \[65-97\]   | 22.3% (192.0/862.0) \[20-25\]   | 24.0% (212.0/885.0) \[21-27\]   |
| Overall            | 73.2% (344.0/470.0) \[69-77\] | 21.2% (367.0/1,731.0) \[19-23\] | 32.3% (711.0/2,201.0) \[30-34\] |

<sup>Created on 2021-09-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

ddsjoberg commented:
hey hey @larmarange ! this all looks great!

To check that I fully understand the benefits here:
1. The function to calculate the custom stats has both full data and subsetted data arguments. This makes it easier for users create their own summary stat functions, because they don't need to deal with the dimensions of n by m data frames (for variable X by variable tables).
2. The table maintains the `tbl_summary` class and structure, which allows all the other functions designed to work with `tbl_summary()` objects will still work, which is fantastic.

I was thinking more something like `tbl_base()` that creates the bare minimum of a gtsummary table (the label column only, and the other needed internals to print properly), and we could add another function similar to `add_stat()` to add the new columns. BUT, your solution I think will be both easier to use AND is the most common customization users need.

Sorry the review took me time to get to! I am still processing/thinking about it all. I think the implementation looks great as it is. Perhaps we could release this in the next version of gtsummary with an experimental tag, indicating there may be breaking changes to it if we decide on a some changes after we've sat with it for a while.

larmarange commented:
> 1. The function to calculate the custom stats has both full data and subsetted data arguments. This makes it easier for users create their own summary stat functions, because they don't need to deal with the dimensions of n by m data frames (for variable X by variable tables).

Exact. The purpose is really to simplify the writing of the function while allowing complex stats.

In addition, as you can see with the examples, the system allows providing helpers (e.g. `ratio_summary()`, `continuous_summary()`, ...) and can be easily extended. It is even possible to reproduce `tbl_continous()` and to mix it with other stats as you can define a different function for each variable.

> 2. The table maintains the `tbl_summary` class and structure, which allows all the other functions designed to work with `tbl_summary()` objects will still work, which is fantastic.

This is one of the advantages. Here we focus on the computation of the stats. For the rest (e.g. formatting the stats, creating the table, the labels, etc.) we use all mechanisms already in place by `tbl_summary()`.

> 
> I was thinking more something like `tbl_base()` that creates the bare minimum of a gtsummary table (the label column only, and the other needed internals to print properly), and we could add another function similar to `add_stat()` to add the new columns. BUT, your solution I think will be both easier to use AND is the most common customization users need.
> 
> Sorry the review took me time to get to! I am still processing/thinking about it all. I think the implementation looks great as it is. Perhaps we could release this in the next version of gtsummary with an experimental tag, indicating there may be breaking changes to it if we decide on some changes after we've sat with it for a while.

I also think that an `experimental` tag would be a good idea. However, there is still some work to be done before release:

- [ ] Improve documentation
- [ ] Add unit tests
- [ ] Add more examples
- [ ] Integrate some helpers presented here in the discussion
- [ ] Add an option for adding an overall raw

There is still some open questions and I would appreciate your feedback on it:

- [ ] It would be nice to improve default value `adding_formatting_as_attr()` when the stat is unknown (checking if the stat is numeric or not and default to `style_number()` or `as.character()`. Should it be done here or in a separate PR?
- [ ] Could we open a separate PR to discuss the opportunity of simplifying the syntax in `.formula_list_to_named_list()`, i.e. allowing to write `~ something` as a shortcut of `everything() ~ something`?

We should also consider adding a `tbl_custom_svysummary()` function doing the same thing but for survey datasets

ddsjoberg commented:
>  It would be nice to improve default value `adding_formatting_as_attr()` when the stat is unknown (checking if the stat is numeric or not and default to `style_number()` or `as.character()`. Should it be done here or in a separate PR?

Let's work on this in a separate PR. I don't recall exactly why, but I think there is an assumption in `inline_text()` that requires some cols to be numeric after conversion....i have a vague recollection that it's related to the specification of `conf.level`?

ddsjoberg commented:
hey @larmarange, a couple of questions:
1. Do you think we need to add these variables to the data frame? They are easily accessible with `data[[variable]]`, right?
![image](https://user-images.githubusercontent.com/26774684/134411517-dd219280-0042-4d3c-ba7d-0fbeec166024.png)
2. We should add a word of caution to the `tbl_custom_summary()` help file about using the associated `tbl_summary()` functions. Something like `add_overall()` works well, but `add_p()` is going to give misleading results.

```r
mean_age <- function(data,...) {
  dplyr::tibble(mean_age = mean(data$age, na.rm = TRUE))
}
mean_age(trial)


trial %>%
  tbl_custom_summary(
    include = c("grade", "response", "stage"),
    by = "trt",
    stat_fns = everything() ~ mean_age,
    statistic = everything() ~ "{mean_age}",
    digits = everything() ~ 1
  ) %>%
  add_p() %>%
  add_overall() %>%
  modify_footnote(all_stat_cols() ~ "Mean age") %>%
  as_kable()
```
|**Characteristic** |**Overall**, N = 200 |**Drug A**, N = 98 |**Drug B**, N = 102 |**p-value** |
|:------------------|:--------------------|:------------------|:-------------------|:-----------|
|Grade              |                     |                   |                    |0.9         |
|I                  |46.2                 |45.9               |46.4                |            |
|II                 |47.5                 |44.6               |50.3                |            |
|III                |48.1                 |51.0               |45.7                |            |
|Tumor Response     |49.8                 |50.1               |49.5                |0.5         |
|Unknown            |7                    |3                  |4                   |            |
|T Stage            |                     |                   |                    |0.9         |
|T1                 |46.8                 |44.1               |49.5                |            |
|T2                 |48.1                 |50.2               |46.4                |            |
|T3                 |49.4                 |48.8               |50.0                |            |
|T4                 |44.8                 |45.3               |44.3                |            |

ddsjoberg commented:
hi @larmarange !

How are the data subset with a continuous variable? I expected the means below to be the same?

``` r
library(gtsummary)

mean_age <- function(data,...) {
  dplyr::tibble(mean_age = mean(data$age, na.rm = TRUE))
}
trial %>%
  tbl_custom_summary(
    include = "marker",
    by = "trt",
    stat_fns = everything() ~ mean_age,
    statistic = everything() ~ "{mean_age}",
    digits = everything() ~ 1,
    missing = "no"
  ) %>%
  add_overall(last = TRUE) %>%
  as_kable()
```

| **Characteristic**   | **Drug A**, N = 98 | **Drug B**, N = 102 | **Overall**, N = 200 |
|:---------------------|:-------------------|:--------------------|:---------------------|
| Marker Level (ng/mL) | 46.5               | 47.5                | 47.0                 |

``` r
trial %>%
  select(trt, age) %>%
  dplyr::group_by(trt) %>%
  dplyr::summarise(
    mean_age = mean(age, na.rm = TRUE)
  )
#> # A tibble: 2 x 2
#>   trt    mean_age
#>   <chr>     <dbl>
#> 1 Drug A     47.0
#> 2 Drug B     47.4
```

<sup>Created on 2021-09-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

larmarange commented:
> 1. Do you think we need to add these variables to the data frame? They are easily accessible with `data[[variable]]`, right?
>    ![image](https://user-images.githubusercontent.com/26774684/134411517-dd219280-0042-4d3c-ba7d-0fbeec166024.png)

I agree that it may be unnecessary for users. I have removed their mention from the doc.

However, they may remain useful internally. For example in this part of code:

```r
  if (!is.null(dichotomous_value))
    data <- data %>%
    mutate(
      # adding dichotomous level (in case it is unobserved)
      .variable = forcats::fct_expand(
        as.factor(.data$.variable),
        as.character(dichotomous_value)
      )
    )
```

Here `data$.variable` is transformed but `data[[variable]]` remain unchanged.

larmarange commented:
> 2\. We should add a word of caution to the `tbl_custom_summary()` help file about using the associated `tbl_summary()` functions. Something like `add_overall()` works well, but `add_p()` is going to give misleading results.

I'm updating the doc.

larmarange commented:
> How are the data subset with a continuous variable? I expected the means below to be the same?

For continuous variables, the data subset removes `NA` values. If you want to display all rows, there is now an option `overall_row`.

``` r
library(gtsummary)
#> #BlackLivesMatter
mean_age <- function(data,...) {
  dplyr::tibble(mean_age = mean(data$age, na.rm = TRUE))
}

trial %>%
  tbl_custom_summary(
    include = "marker",
    by = "trt",
    stat_fns = everything() ~ mean_age,
    statistic = everything() ~ "{mean_age}",
    digits = everything() ~ 1
  ) %>%
  add_overall(last = TRUE) %>%
  as_kable()
```

| **Characteristic**   | **Drug A**, N = 98 | **Drug B**, N = 102 | **Overall**, N = 200 |
|:---------------------|:-------------------|:--------------------|:---------------------|
| Marker Level (ng/mL) | 46.5               | 47.5                | 47.0                 |
| Unknown              | 6                  | 4                   | 10                   |

``` r
trial %>%
  dplyr::group_by(trt, is.na(marker)) %>%
  dplyr::summarise(
    mean_age = mean(age, na.rm = TRUE)
  )
#> `summarise()` has grouped output by 'trt'. You can override using the `.groups` argument.
#> # A tibble: 4 x 3
#> # Groups:   trt [2]
#>   trt    `is.na(marker)` mean_age
#>   <chr>  <lgl>              <dbl>
#> 1 Drug A FALSE               46.5
#> 2 Drug A TRUE                54.5
#> 3 Drug B FALSE               47.5
#> 4 Drug B TRUE                47
trial %>%
  tbl_custom_summary(
    include = NULL,
    by = "trt",
    stat_fns = everything() ~ mean_age,
    statistic = everything() ~ "{mean_age}",
    digits = everything() ~ 1,
    overall_row = TRUE
  ) %>%
  add_overall(last = TRUE) %>%
  as_kable()
```

| **Characteristic** | **Drug A**, N = 98 | **Drug B**, N = 102 | **Overall**, N = 200 |
|:-------------------|:-------------------|:--------------------|:---------------------|
| Overall            | 47.0               | 47.4                | 47.2                 |

``` r
trial %>%
  dplyr::group_by(trt) %>%
  dplyr::summarise(
    mean_age = mean(age, na.rm = TRUE)
  )
#> # A tibble: 2 x 2
#>   trt    mean_age
#>   <chr>     <dbl>
#> 1 Drug A     47.0
#> 2 Drug B     47.4
```

<sup>Created on 2021-09-23 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

ddsjoberg commented:
> For continuous variables, the data subset removes NA values. If you want to display all rows, there is now an option overall_row.
 Ahh, yes! That makes sense and is expected!

ddsjoberg commented:
Regarding `.variable` and `.by`...I am not yet 100% sold that we need them in the data set. One issue is that names will now be protected. If someone had these variables in their data set, they would experience unexpected results. Also, I am not sure a user need a version of the variable they can modify and a version that remains the same....if that were needed in a specific case, they could make a copy of the variable within their small stat_fun?

ddsjoberg commented:
@larmarange hello!

How do you want to handle unobserved levels of a factor level? At the moment, the unobserved level is ignored, while in `tbl_summary()` the unobserved level is shown.

``` r
library(gtsummary)

mean_age <- function(data, ...) {
  dplyr::tibble(mean_age = mean(data$age, na.rm = TRUE))
}

trial %>%
  # adding an unobserved level in grade
  mutate(grade = forcats::fct_expand(grade, "IV")) %>%
  tbl_custom_summary(
    include = grade,
    by = "trt",
    stat_fns = everything() ~ mean_age,
    statistic = everything() ~ "{mean_age}",
    digits = everything() ~ 1
  ) %>%
  as_kable()
```

| **Characteristic** | **Drug A**, N = 98 | **Drug B**, N = 102 |
|:-------------------|:-------------------|:--------------------|
| Grade              |                    |                     |
| I                  | 45.9               | 46.4                |
| II                 | 44.6               | 50.3                |
| III                | 51.0               | 45.7                |

<sup>Created on 2021-09-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

larmarange commented:
> How do you want to handle unobserved levels of a factor level? At the moment, the unobserved level is ignored, while in `tbl_summary()` the unobserved level is shown.

Nice catch. I have added a `.drop = FALSE` in `dplyr::group_by()`

``` r
library(gtsummary)

df <- trial %>%
  mutate(grade = forcats::fct_expand(grade, "IV"))

df %>%
  tbl_summary(
    include = "grade",
    by = "trt",
    percent = "row"
  ) %>%
  as_kable()
```

| **Characteristic** | **Drug A**, N = 98 | **Drug B**, N = 102 |
|:-------------------|:-------------------|:--------------------|
| Grade              |                    |                     |
| I                  | 35 (51%)           | 33 (49%)            |
| II                 | 32 (47%)           | 36 (53%)            |
| III                | 31 (48%)           | 33 (52%)            |
| IV                 | 0 (NA%)            | 0 (NA%)             |

``` r
df %>%
  tbl_custom_summary(
    include = "grade",
    by = "trt",
    stat_fns = everything() ~ continuous_summary("age"),
    statistic = everything() ~ "{median} [{p25}-{p75}]",
    overall_row = TRUE
  ) %>%
  as_kable()
```

| **Characteristic** | **Drug A**, N = 98 | **Drug B**, N = 102 |
|:-------------------|:-------------------|:--------------------|
| Overall            | 46 \[37-59\]       | 48 \[39-56\]        |
| Grade              |                    |                     |
| I                  | 46 \[36-60\]       | 48 \[42-55\]        |
| II                 | 44 \[31-54\]       | 50 \[43-57\]        |
| III                | 52 \[42-60\]       | 45 \[36-52\]        |
| IV                 | NA \[NA-NA\]       | NA \[NA-NA\]        |

<sup>Created on 2021-09-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

larmarange commented:
> Regarding `.variable` and `.by`...I am not yet 100% sold that we need them in the data set. One issue is that names will now be protected. If someone had these variables in their data set, they would experience unexpected results. Also, I am not sure a user need a version of the variable they can modify and a version that remains the same....if that were needed in a specific case, they could make a copy of the variable within their small stat_fun?

I will remove the `.by` and the `.variable`.

You just need to be aware that, for dichotomous variables, they all be transformed into factors, i.e. an integer 0/1 variable will be seen in the data received by `stat_fns` as a factor with levels "0" and "1".

See this code in `summarize_custom()`:

```r
  if (!is.null(dichotomous_value))
    data <- data %>%
    mutate(
      # adding dichotomous level (in case it is unobserved)
      .variable = forcats::fct_expand(
        as.factor(.data$.variable),
        as.character(dichotomous_value)
      )
    )
```

But this is probably not a concern. Just something to remember.

I will update the code later today (sorry but very busy with my regular work right now)

larmarange commented:
Hi @ddsjoberg 

I have added helpers and more examples. The code development is almost ready.

How many unit tests do you want me to add? Due to the time of all tests, do you have any policy? Should I use `skip_on_cran()`?

Should I update NEWS?

Any other request or check list?

ddsjoberg commented:
> I have added helpers and more examples. The code development is almost ready.

Amazing!

> How many unit tests do you want me to add? Due to the time of all tests, do you have any policy? Should I use `skip_on_cran()`?

Add however many you feel is good. I previously wrote most unit tests to check for errors, but I now add `expect_equal()` to each function/argument to ensure the results are what we expect. Yes, please add `skip_on_cran()` at the top. The package has become so large and we have to skip on CRAN so the checks don't time out and fail.

> Should I update NEWS?

Can you add the bullet points that will be added to NEWS.md to the first comment in the pull request?  After I do the full review, I'll add those items to the NEWS file, increment the version number, and merge the PR.  I do this as a last step because we often have multiple PRs open and if each of them update the NEWS we encounter merge conflicts.

> Any other request or check list?

I think that is everything! I am excited! Thank you for adding this function. When it's ready, I think I'll re-write `tbl_continuous()` to simplify the function immensely! 
@larmarange

larmarange commented:
Dear @ddsjoberg 

I have updated the documentation, performed a `spell_check()`, and added several unit tests.

I guess that the PR is now ready for review.

Cheers

ddsjoberg commented:
Amazing @larmarange ! I'll take a look!

ddsjoberg reviewed (approved):
@larmarange this all looks great!

larmarange commented:
Thanks @ddsjoberg 

Should we wait a little and additional feedback before working on `tbl_custom_svysummary()` ?

ddsjoberg commented:
> Should we wait a little and additional feedback before working on `tbl_custom_svysummary()` ?

Yeah, I think that is a good idea!
